From d82f542232c7ead4b1d743f3c4252c98dad4e7c4 Mon Sep 17 00:00:00 2001
From: ChonDoit <thphantomblog@gmail.com>
Date: Tue, 11 Aug 2026 09:06:11 +0000
Subject: [PATCH] Fix brightness slider curve for some devices

Some devices report max brightness as 2047 or 4095

[Port to A16] SystemProperties import already present upstream (sDisableGammaConversion); the two conversion functions and header constants patched identically.

Change-Id: Iecfc09708db045a47b3bb1d96633512179180e87
---
 .../settingslib/display/BrightnessUtils.java       | 14 +++++++++-----
 1 file changed, 9 insertions(+), 5 deletions(-)

diff --git a/packages/SettingsLib/DisplayUtils/src/com/android/settingslib/display/BrightnessUtils.java b/packages/SettingsLib/DisplayUtils/src/com/android/settingslib/display/BrightnessUtils.java
index 3c06f8ae9972..ceecd76ddd81 100644
--- a/packages/SettingsLib/DisplayUtils/src/com/android/settingslib/display/BrightnessUtils.java
+++ b/packages/SettingsLib/DisplayUtils/src/com/android/settingslib/display/BrightnessUtils.java
@@ -19,14 +19,18 @@ package com.android.settingslib.display;
 import android.os.SystemProperties;
 import android.util.MathUtils;
 
+import com.android.internal.display.BrightnessSynchronizer;
+
 /** Utility methods for calculating the display brightness. */
 public class BrightnessUtils {
 
     public static final boolean sDisableGammaConversion =
             SystemProperties.getBoolean("sys.brightness.disable_gamma_conversion", false);
 
+    public static final boolean LowGammaBrightness = Boolean.parseBoolean(SystemProperties.get("persist.sys.phh.low_gamma_brightness", "false"));
+
     public static final int GAMMA_SPACE_MIN = 0;
-    public static final int GAMMA_SPACE_MAX = 65535;
+    public static final int GAMMA_SPACE_MAX = LowGammaBrightness ? 255 : 65535;
 
     // Hybrid Log Gamma constant values
     private static final float R = 0.5f;
@@ -102,9 +106,8 @@ public class BrightnessUtils {
         // it shouldn't be out of bounds.
         final float normalizedRet = MathUtils.constrain(ret, 0, 12);
 
-        // Re-normalize to the range [0, 1]
-        // in order to derive the correct setting value.
-        return MathUtils.lerp(min, max, normalizedRet / 12);
+        return LowGammaBrightness ? MathUtils.constrain(BrightnessSynchronizer.brightnessIntToFloat(val),
+                         min, max) : MathUtils.lerp(min, max, normalizedRet / 12);
     }
 
     /**
@@ -156,6 +159,7 @@ public class BrightnessUtils {
             ret = A * MathUtils.log(normalizedVal - B) + C;
         }
 
-        return Math.round(MathUtils.lerp(GAMMA_SPACE_MIN, GAMMA_SPACE_MAX, ret));
+        return LowGammaBrightness ? BrightnessSynchronizer.brightnessFloatToInt(
+                       MathUtils.constrain(val, min, max)) : Math.round(MathUtils.lerp(GAMMA_SPACE_MIN, GAMMA_SPACE_MAX, ret));
     }
 }
-- 
2.43.0

